src/actions/plugins/imswitcher.ts

```diff
@@ -0,0 +1,137 @@
+import { ModeName } from '../../mode/mode';
+import { existsSync } from 'fs';
+import { configuration } from '../../configuration/configuration';
+import * as util from '../../util/util';
+import { Message } from '../../util/message';
+
+// InputMethodSwitcher change input method automatically when mode changed
+export class InputMethodSwitcher {
+  private savedIMKey = '';
+
+  public async switchInputMethod(prevMode: ModeName, newMode: ModeName) {
+    if (configuration.autoSwitchInputMethod.enable !== true) {
+      return;
+    }
+    if (!this.validateConfiguration()) {
+      this.disableIMSwitch();
+      return;
+    }
+    // when you exit from insert-like mode, save origin input method and set it to default
+    if (
+      prevMode === ModeName.Insert ||
+      prevMode === ModeName.SurroundInputMode ||
+      prevMode === ModeName.Replace
+    ) {
+      if (
+        newMode !== ModeName.Insert &&
+        newMode !== ModeName.SurroundInputMode &&
+        newMode !== ModeName.Replace
+      ) {
+        this.switchToDefaultIM();
+        return;
+      }
+    }
+    // when you enter insert-like mode, resume origin input method
+    if (
+      newMode === ModeName.Insert ||
+      newMode === ModeName.SurroundInputMode ||
+      newMode === ModeName.Replace
+    ) {
+      if (
+        prevMode !== ModeName.Insert &&
+        prevMode !== ModeName.SurroundInputMode &&
+        prevMode !== ModeName.Replace
+      ) {
+        this.resumeIM();
+      }
+    }
+  }
+
+  // save origin input method and set input method to default
+  private async switchToDefaultIM() {
+    const obtainIMCmd = configuration.autoSwitchInputMethod.obtainIMCmd;
+    const rawObtainIMCmd = this.getRawCmd(obtainIMCmd);
+    if (existsSync(rawObtainIMCmd)) {
+      try {
+        const insertIMKey = await util.executeShell(obtainIMCmd);
+        if (insertIMKey !== undefined) {
+          this.savedIMKey = insertIMKey.trim();
+        }
+      } catch (error) {
+        console.log(error);
```

Comment:
Can you use the logger? https://github.com/VSCodeVim/Vim/blob/master/src/mode/modeHandler.ts#L335

Reply:
OK